### PR TITLE
UI Texture Atlas support

### PR DESCRIPTION
**Pull request by ManevilleF**
# Objective

Adds support for texture atlas in `bevy_ui` Nodes

## Solution

- I created a custom `UiTextureAtlas` component with a simple texture `index` and a `Handle<bevy_sprite::TextureAtlas>`
- I plugged a new extraction system in the render pipeline
- I added new UI Bundles using `UiTextureAtlas`
- I added an example
- I added an extra system for the `ImageMode` support

### CODE REVIEW

**Comment by ManevilleF** [comment]
I don't know why but it doesn't work :( 

`thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', crates/bevy_ui/src/render/mod.rs:512:70`

It seems that the `TextureAtlas::texture`(a `Handle<Image>`) is not set in `RenderAssets<Image>`

**Comment by ManevilleF** [comment]
Fixed, everything seems to be working correctly

**Comment by ManevilleF** [comment]
The failing test is a false negative fixed by #3789 I think

File: crates/bevy_ui/src/entity.rs [hunk 1]
```diff
@@ -54,6 +55,29 @@ pub struct ImageBundle {
     pub visibility: Visibility,
 }
 
+/// A UI node that is an image with texture sheet
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/entity.rs:58]:
```suggestion
/// A UI node that displays an image stored in a texture sheet
```

File: crates/bevy_ui/src/entity.rs [hunk 2]
```diff
@@ -54,6 +55,29 @@ pub struct ImageBundle {
...
+pub struct ImageSheetBundle {
+    /// Describes the size of the node
+    pub node: Node,
+    /// Describes the style including flexbox settings
+    pub style: Style,
+    /// Configures how the image should scale
+    pub image_mode: ImageMode,
+    /// The calculated size based on the given image
+    pub calculated_size: CalculatedSize,
+    /// The color of the node
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/entity.rs:69]: This should describe how the blending is done.

File: crates/bevy_ui/src/entity.rs [hunk 3]
```diff
@@ -132,6 +156,48 @@ impl Default for ButtonBundle {
     }
 }
 
+/// A UI node that is a button with a texture sheet
+#[derive(Bundle, Clone, Debug)]
+pub struct ButtonSheetBundle {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/entity.rs:161]: I don't like the creation of a dedicated bundle here: this should be done compositionally.

File: crates/bevy_ui/src/entity.rs [hunk 4]
```diff
@@ -132,6 +156,48 @@ impl Default for ButtonBundle {
...
+    pub texture_atlas: UiTextureAtlas,
+    /// The transform of the node
+    pub transform: Transform,
+    /// The global transform of the node
+    pub global_transform: GlobalTransform,
+    /// Describes the visibility properties of the node
+    pub visibility: Visibility,
+}
+
+impl Default for ButtonSheetBundle {
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/entity.rs:184]: Shouldn't the derive work fine here?

File: crates/bevy_ui/src/render/mod.rs [hunk 5]
```diff
@@ -82,12 +82,22 @@ pub fn build_ui_render(app: &mut App) {
         .add_system_to_stage(RenderStage::Extract, extract_ui_camera_phases)
         .add_system_to_stage(
             RenderStage::Extract,
+            // This system is the first of `RenderStage::Extract` it is responsible for both:
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:85]: This should be on doc strings for extract_uinodes` instead.

File: examples/ui/button_with_atlas.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
+use bevy::prelude::*;
+
+/// This example illustrates how to create a button that changes color and text based on its
```
Comment (alice-i-cecile) [inline examples/ui/button_with_atlas.rs:3]: This should be a module-style comment: see #4438.

File: examples/README.md [hunk 7]
```diff
@@ -243,6 +243,7 @@ Example | File | Description
 Example | File | Description
 --- | --- | ---
 `button` | [`ui/button.rs`](./ui/button.rs) | Illustrates creating and updating a button
+`button_with_atlas` | [`ui/button_with_atlas.rs`](./ui/button_with_atlas.rs) | Illustrates creating and updating a button with a texture atlas
```
Comment (alice-i-cecile) [inline examples/README.md:246]: I'd prefer if we could roll this into the `button.rs` example, and just create two different buttons there.

**Comment by ManevilleF** [comment]
Thanks for the review, I completely forgot about this PR.

I think I'll redo it completely from the `main` branch following your remarks.

About the use cases I see both:
- Animated buttons (and anything really)
- Global texture mapping for UI

**Comment by ManevilleF** [comment]
@alice-i-cecile I'm closing this in favor of https://github.com/bevyengine/bevy/pull/5070

**Pull request closed without merging**